src/rdsk.c

```diff
@@ -27,473 +27,800 @@
  ** @date: 15Oct10, petros@petroskoutoupis.com
  ********************************************************************************/
 
-#include "common.h"
-#include "cli.h"
+#include "rdsk.h"
+#include "utils.h"
+
+#ifdef SERVER
+#include "rapiddiskd.h"
+#endif
 #include <malloc.h>
 #include <linux/fs.h>
 #include <sys/ioctl.h>
+#include <libdevmapper.h>
 
 #define BYTES_PER_BLOCK		512
 #define IOCTL_RD_GET_STATS	0x0529
 #define IOCTL_RD_BLKFLSBUF	0x0531
 
-struct RD_PROFILE *rdsk_head =  (struct RD_PROFILE *) NULL;
-struct RD_PROFILE *rdsk_end =   (struct RD_PROFILE *) NULL;
-struct RC_PROFILE *cache_head = (struct RC_PROFILE *) NULL;
-struct RC_PROFILE *cache_end =  (struct RC_PROFILE *) NULL;
+struct RD_PROFILE *rdsk_head = NULL;
+struct RD_PROFILE *rdsk_end = NULL;
+struct RC_PROFILE *cache_head = NULL;
+struct RC_PROFILE *cache_end = NULL;
+
+static const size_t RC_STATS_OFFSETS[] = {
+		offsetof(RC_STATS, device),
+		offsetof(RC_STATS, reads),
+		offsetof(RC_STATS, writes),
+		offsetof(RC_STATS, cache_hits),
+		offsetof(RC_STATS, replacement),
+		offsetof(RC_STATS, write_replacement),
+		offsetof(RC_STATS, read_invalidates),
+		offsetof(RC_STATS, write_invalidates),
+		offsetof(RC_STATS, uncached_reads),
+		offsetof(RC_STATS, uncached_writes),
+		offsetof(RC_STATS, disk_reads),
+		offsetof(RC_STATS, disk_writes),
+		offsetof(RC_STATS, cache_reads),
+		offsetof(RC_STATS, cache_writes),
+
+		/* Unsupported in this release */
+		offsetof(RC_STATS, read_ops),
+		offsetof(RC_STATS, write_ops),
+};
+
+static const size_t WC_STATS_OFFSETS[] =  {
+		offsetof(WC_STATS, device),
+		offsetof(WC_STATS, expanded),
+		offsetof(WC_STATS, errors),
+		offsetof(WC_STATS, num_blocks),
+		offsetof(WC_STATS, num_free_blocks),
+		offsetof(WC_STATS, num_wb_blocks),
+		/* For 5.15 and later */
+		offsetof(WC_STATS, num_read_req),
+		offsetof(WC_STATS, num_read_cache_hits),
+		offsetof(WC_STATS, num_write_req),
+		offsetof(WC_STATS, num_write_uncommitted_blk_hits),
+		offsetof(WC_STATS, num_write_committed_blk_hits),
+		offsetof(WC_STATS, num_write_cache_bypass),
+		offsetof(WC_STATS, num_write_cache_alloc),
+		offsetof(WC_STATS, num_write_freelist_blocked),
+		offsetof(WC_STATS, num_flush_req),
+		offsetof(WC_STATS, num_discard_req),
+};
+
+/**
+ * It's a printf() function that can optionally return the formatted string instead of printing it
+ *
+ * @param format_string The error format string.
+ * @param return_message This is a pointer to a string that will be filled with the error message. If this is NULL, the
+ * error message will be printed to stdout.
+ * @param ... Values to be inserted.
+ */
+void print_error(char *format_string, char *return_message, ...) {
```

Comment:
We can leave it as print_error but seeing how it doesn't just print errors, would it make more sense to rename? If not, it is okay, we can keep it as print_error.

Reply:
You're right, it's not a good name at all... any suggestion?